cmd/podman/images/scp.go

```diff
@@ -0,0 +1,317 @@
+	"github.com/containers/podman/v3/cmd/podman/system/connection"
+	"github.com/containers/podman/v3/pkg/domain/entities"
+	scpD "github.com/dtylman/scp"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"github.com/spf13/cobra"
+	"golang.org/x/crypto/ssh"
+)
+
+var (
+	saveScpDescription = `Securely copy an image from one host to another.`
+	imageScpCommand    = &cobra.Command{
+		Use:   "scp [options] IMAGE [IMAGE...]",
+		Long:  saveScpDescription,
+		Short: "securely copy images",
+		RunE:  scp,
+		Args: func(cmd *cobra.Command, args []string) error {
+			if len(args) == 0 {
+				return errors.Errorf("need at least 1 argument")
+			}
+			return nil
+		},
+		ValidArgsFunction: common.AutocompleteImages,
+		Example:           `podman image scp myimage:latest otherhost::`,
+	}
+)
+
+var (
+	scpOpts entities.ImageScpOptions
+)
+
+func init() {
+	registry.Commands = append(registry.Commands, registry.CliCommand{
+		Command: imageScpCommand,
+		Parent:  imageCmd,
+	})
+	scpFlags(imageScpCommand)
+}
+
+func scpFlags(cmd *cobra.Command) {
+	flags := cmd.Flags()
+	flags.BoolVarP(&scpOpts.Save.Quiet, "quiet", "q", false, "Suppress the output")
+}
+
+func scp(cmd *cobra.Command, args []string) (finalErr error) {
+	var (
+		tags []string
+		err  error
+	)
+	if scpOpts.Save.Quiet { // set quiet for both load and save
+		scpOpts.Load.Quiet = true
+	}
+	f, err := ioutil.TempFile("", "temp") // open temp dir for load/save output
+	if err != nil {
+		return err
+	}
+
+	scpOpts.Save.Output = f.Name()
+	scpOpts.Load.Input = scpOpts.Save.Output
+	if err := parse.ValidateFileName(saveOpts.Output); err != nil {
+		return err
+	}
+	confR, err := config.NewConfig("") // create a hand made config for the remote engine since we might use remote and native at once
+	if err != nil {
+		return errors.Wrapf(err, "could not make config")
+	}
+	abiEng, err := registry.NewImageEngine(cmd, args) // abi native engine
+	if err != nil {
+		return err
+	}
+
+	cfg, err := config.ReadCustomConfig() // get ready to set ssh destination if necessary
+	if err != nil {
+		return err
+	}
+	urls, iden, serv := parseLocation(args, cfg) // parses connection data and "which way" we are loading and saving
+
+	// TODO: Add podman remote support
+
+	uri := []*urlP.URL{}
+	for i, url := range urls {
+		urlt, err := urlP.Parse(url) // create an actual url to pass to exec command
+		if err != nil {
+			return err
+		}
+		uri = append(uri, urlt)
+		if uri[i].User.Username() == "" {
+			if uri[i].User, err = connection.GetUserInfo(uri[i]); err != nil {
+				return err
+			}
+		}
+	}
+
+	confR.Engine = config.EngineConfig{Remote: true, CgroupManager: "cgroupfs", ServiceDestinations: serv} // pass the service dest (either remote or something else) to engine
+
+	// TODO: add podman remote support
+
+	switch {
+	case scpOpts.FromRemote: // if we want to load FROM the remote
+		_, err := execCommand(iden[0], uri[0], scpOpts.Save.Output, scpOpts.ImageName, args[0], "", false)
```

Comment:
Not to hold this up, but I believe we should be able to do this without execing a separate process.  We should be able to just create a go func that does the remote code.

Reply:
@rhatdan `execCommand()` is a function w/in scp.go that just uses the input given by users to send the proper information over the ssh connection before executing the command on the remote machine.

Reply:
currently, as `system connection add` does, the best way to execute a command on a remote machine is to do a `session.Run()` on the remote machine. We could experiment down the line with being less explicit an smarter with how we execute these commands but for now it seems pretty straightforward.

Reply:
Explanation of `execCommand()`:

This function (written by me) validates the connection given by the user, and then dials and connects to the host. After this, depending on if we are saving or loading remotely, I execute the followig:

```
podman := "podman"
	if v, found := os.LookupEnv("PODMAN_BINARY"); found {
		podman = v // get path to podman bin
	}
	run := podman + " image save " + imageFileOut + " --output=" + imageFileIn // run ssh image load of the file copied via scp. Files are reverse in thie case...
	var buffer bytes.Buffer
	session.Stdout = &buffer

	if err := session.Run(run); err != nil { // run the command declared above
		return "", err
	}
	n, err := scpD.CopyFrom(dial, imageFileIn, imageFileIn)
	if err != nil {
		fmt.Print(n, " bytes copied before err")
		return "", err
	}
	return "", nil
```

This uses the ssh session to exeucte a podman command on the remote client. The above snipped is the load portion, there is an equivalent one for save.

Reply:
I.e., to be explicit, this does not use the remote API at all, just (the equivalent of?) `scp` to copy files around, and the CLI.

(I don’t know enough to have an opinion on whether this is preferable, undesirable, or not important.)

Reply:
resolved on IRC, deciding to keep execCommand as is for now